signac/core/collection_api.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from contextlib import contextmanager
+from abc import abstractmethod
+from abc import ABCMeta
+from collections import defaultdict
+from collections.abc import Collection
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+
+class SyncedCollectionABCMeta(ABCMeta):
+    """ Metaclass for the definition of SyncedCollection.
+
+    This metaclass automatically registers synced data structures' definition,
+    this is used when recursively converting synced data structures to determine
+    what to convert their children into.
+    """
+    def __init__(cls, name, bases, dct):
+        if not hasattr(cls, 'registry'):
+            cls.registry = defaultdict(list)
+        else:
+            if not cls.__abstractmethods__:
+                cls.registry[cls.backend].append(cls)
+        return super().__init__(name, bases, dct)
+
+
+class SyncedCollection(Collection, metaclass=SyncedCollectionABCMeta):
+    """The base synced collection represents a collection that is synced with a backend.
+
+    The class is intended for use as an ABC. In addition, it declares abstract
+    methods that must be implemented by any subclass.The SyncedCollection is a
```

Comment:
```suggestion
    The class is intended for use as an ABC. In addition, it declares abstract
    methods that must be implemented by any subclass. The SyncedCollection is a
```

Reply:
> In addition, it declares abstract methods [...]

Isn't that the definition of an abstract base class? That it defines abstract methods? Sounds like a tautology to me

Reply:
I will correct that.